File: ToolsX/leetcode/0007/0007.py
class Solution:
    def reverse(self, x: int) -> int:
        """

        要注意的是 32 位整型，如果反转可能溢出， 应该用 long 接收
        一定要注意符号，因为整除是得到 <= 商的整数，所以
        -12//10=-2
        所以
        -12%10=8
        但是在 java 里却是 -2
        >>> Solution().reverse(123)
        321
        >>> Solution().reverse(-123)
        -321
        >>> Solution().reverse(0)
        0
        >>> Solution().reverse(1)
        1
        >>> Solution().reverse(1534236469)
        0
        """
        if x < 0:
            symbol = -1
            x = -x
        else:
            symbol = 1
        r = 0
        while x:
            r *= 10
            r += x % 10
            x //= 10
        if symbol < 0:
            # 最大可取到 1<<31
            if r > 1 << 31:
                return 0
        else:
            # 最大可取到 (1<<31)-1
            if r >= 1 << 31:
                return 0
        return r * symbol


if __name__ == '__main__':
    import doctest

    doctest.testmod(verbose=True)

File: ToolsX/leetcode/0007/0007.java
/**
 * @author pingfangx
 * @date 2019/7/7
 */
class Solution {
    /**
     * 这个这么快么，可是耗时多，可能是不应该用 long 去存
     * Runtime: 1 ms, faster than 100.00% of Java online submissions for Reverse Integer.
     * Memory Usage: 33.4 MB, less than 7.89% of Java online submissions for Reverse Integer.
     */
    public int reverse(int x) {
        long r = 0;
        int symbol;
        if (x < 0) {
            x = -x;
            symbol = -1;
        } else {
            symbol = 1;
        }
        while (x > 0) {
            r *= 10;
            r += x % 10;
            x /= 10;
        }
        r *= symbol;
        if (r < Integer.MIN_VALUE || r > Integer.MAX_VALUE) {
            return 0;
        }
        return (int) r;
    }
}
